rubicon/objc/runtime.py

```diff
@@ -981,3 +990,23 @@ def set_ivar(obj, varname, value, weak=False):
         libobjc.object_setIvar(obj, ivar, value)
     else:
         memmove(obj.value + libobjc.ivar_getOffset(ivar), addressof(value), sizeof(vartype))
+
+
+class autoreleasepool:
+    """
+    A context manager that has the same effect as a @autoreleasepool block in Objective-C.
+
+    Any objects that are autoreleased within the context will receive a release message when exiting the context. When
+    running an event loop, AppKit will create an autorelease pool at the beginning of each cycle of the event loop and
+    drain it at the end. You therefore do not need to use @autoreleasepool blocks when running an event loop. However,
+    they may be still be useful when your code temporarily allocates large amounts of memory which you want to
+    explicitly free before the end of a cycle.
+    """
+    def __init__(self):
+        self.pool = None
+
+    def __enter__(self):
+        self.pool = libobjc.objc_autoreleasePoolPush()
```

Comment:
Speaking of error checking, it might be good to raise an exception here if `self.pool` has already been set. This would guard against reusing the same `autoreleasepool` object in more than one `with` block. (This is very unlikely to happen by accident, but better to check anyway, to avoid weird runtime errors.)

Reply:
Those are both good points. I am considering rewriting the context manager as function decorated with `@contextmanager`. This will automatically ensure matching `__enter__()` and `__exit__()` calls without manual checks, however with some amount of "magic" going on behind the scenes and somewhat enigmatic error messages when the context manager is used improperly.

Reply:
Yeah, using `@contextlib.contextmanager` might be a good idea if it checks this automatically. The less precise error messages aren't a big problem I think, because this error can only show up if you write really unusual code (you'd have to store the `autoreleasepool` object somewhere rather than using the normal `with autoreleasepool(): ...` usage pattern). If it ends up being a problem for some reason, we can always switch back to a manual context manager later.

Reply:
Done.